src/backend/utils/fmgr/fmgr.c

```diff
@@ -773,12 +774,13 @@ fmgr_security_definer(PG_FUNCTION_ARGS)
 						GUC_ACTION_SAVE);
 	}
 
-	if (set_sql_dialect)
+	if (set_sql_dialect && IsTransactionState())
```

Comment:
we should add a test case for this

Reply:
That's something I got stucked for now, jdbc didn't allow to change client_encoding during holding the connection, so can't add test case in jdbc test case, and in engine code we can add a case into install check, but we didn't run install check in Github

Reply:
how about ODBC test?